### DISCUSSION TITLE
Android example app

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#283] [Show and tell]
Implement a very basic Java application using `whisper.cpp`. It can be used as an example for running Whisper on Android.

The [ggwave-java](https://github.com/ggerganov/ggwave-java) project can be used as a good starting point. It already provides the audio capture functionality. Instead of passing it to `ggwave`, we just need to pass it to `whisper.cpp`.

Edit:
Looking for volunteers to help with this - ideally, we would like to have the same functionality demonstrated as in the [iOS example application](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc).

---

**Update:**

Thanks to @Digipom - this is now available!
Checkout the [examples/whisper.android](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.android) example

<img width="300" alt="image" src="https://user-images.githubusercontent.com/1991296/208154256-82d972dc-221b-48c4-bfcb-36ce68602f93.png">

**Comment by Hannes1** [comment]
https://github.com/alphacep/vosk-android-demo 

Maybe could also have a look at vosk demo, it's more adapted for streaming though but just for example...

This is a bit off topic, I've been playing around with react native and wenet also, haven't done IOS yet but yeah the android part is working:
https://github.com/Hannes1/react-native-wenet

**Comment by luke-jr** [comment]
Maybe one of the open source Android keyboard projects would be a good candidate for integration :)

**Comment by CristhianNY** [comment]
it's possible to hear the Spanish language?  , I have been trying but i speak in Spanish but I'm no able to transcribe  spanish

**Reply by ArthurPeabody** [reply]
me, too.

**Comment by rpdrewes** [comment]
I use whisper.cpp from my de-googled Android phone daily for months now for dictation to SMS and email and web search. My preferred method is to use the Konele app as the voice input frontend which sends the audio to my own private server which is running a very simple python wrapper around whisper.cpp. Here is my project: https://github.com/rpdrewes/whisper-websocket-server

Another alternative is this port of whisper.cpp to run natively on Android. I have tested it and it also works well but it is not as fast as my server-based approach in my environment (but if you have a slower server and/or a faster phone it might be better for you). It uses Konele as a frontend for the input as well, I believe. See: https://github.com/alex-vt/WhisperInput

**Comment by zhouwg** [comment]
Thanks for the great whisper.cpp although I heard it too late.

I had been finished [a similar PoC with Mozilla's DeepSpeech](https://github.com/cdeos/DeepSpeech) on Android-based phone.

But I'd like to try to [integrate whisper.cpp with KanTV](https://github.com/cdeos/kantv/issues/64) because it's more closer to "real application".

The following snapshot will be a start point and there's a significant performance issue(could [Google's Highway](https://github.com/google/highway) be used to solve performance issue?)

![Screenshot_20240304_183808_com litongjava whisper android java](https://github.com/ggerganov/whisper.cpp/assets/6889919/de7113e4-75cd-4c88-a078-1919f8c02033)

**Comment by zhouwg** [comment]
integrate whisper.cpp to KanTV [step1:](https://github.com/cdeos/kantv/tree/kantv-poc-with-whispercpp)

![1239754793](https://github.com/ggerganov/whisper.cpp/assets/6889919/c02b7616-511b-4922-97cc-89e0680ddc7a)


How to practise/play with [this branch](https://github.com/cdeos/kantv/tree/kantv-poc-with-whispercpp):

adb logcat | grep KANTV
(logs in Java layer / JNI layer / Native layer would be displayed with **same prefix** properly, so it's helpful for troubleshooting and tracking source codes  and performance bottleneck analysis in whisper.cpp)

![Screenshot from 2024-03-05 21-56-25](https://github.com/ggerganov/whisper.cpp/assets/6889919/580600fe-5fe9-43b1-b00c-07bd9a240633)

**Comment by liam-mceneaney** [comment]
I added a streaming example in a fork: https://github.com/liam-mceneaney/androidwhisper.cpp/tree/master

Also my PR: https://github.com/ggerganov/whisper.cpp/pull/1924

![Screenshot_20240304_195942_WhisperCppDemo](https://github.com/ggerganov/whisper.cpp/assets/69867124/adb334a9-9164-4d51-82e7-e658657f1993)

**Reply by zhouwg** [reply]
thanks so much. It's very useful for other Android developer and it would be **very helpful** for my on-going PoC(which just for personal interests/hobbies).

but I couldn't build it and then study your code accordingly because of GFW(I guess it's because of it's highly sensitive period in China now)
![Screenshot from 2024-03-07 22-05-14](https://github.com/ggerganov/whisper.cpp/assets/6889919/356297f0-c499-41f7-9f85-dbd88104bd5a)

btw, could ggerganov create/open a WeChat account(tip:many personal privacy information(include **face identification**) might be required/provided to open a WeChat account  and you might be known we are used to that) so I can buy some coffee which express my sincerely thanks for your great whisper.cpp. I guess you should receive a lot ZANSHANG(aka "赞赏",  or "reward"  ------- similar to Paypal's personal donation) from programmers in China if you have a WeChat account.

**Reply by zhouwg** [reply]
@liam-mceneaney, thanks so much. I got your key-point of code:
```
//streaming attempt:
JNIEXPORT void JNICALL
Java_com_whispercpp_whisper_WhisperLib_00024Companion_fullStreamTranscribe(
        JNIEnv *env, jobject thiz, jlong context_ptr, jint num_threads, jfloatArray audio_data) {
    UNUSED(thiz);
    struct whisper_context *context = (struct whisper_context *) context_ptr;
    jfloat *audio_data_arr = (*env)->GetFloatArrayElements(env, audio_data, NULL);
    const jsize audio_data_length = (*env)->GetArrayLength(env, audio_data);

    // The below adapted from the Objective-C iOS sample
    struct whisper_full_params params = whisper_full_default_params(WHISPER_SAMPLING_GREEDY);
    params.print_realtime = true;
    params.print_progress = false;
    params.print_timestamps = true;
    params.print_special = false;
    params.translate = false;
    params.language = "en";
    params.n_threads = num_threads; //how many threads can I use on an S23?
    //potentially use an initial prompt for custom vocabularies?
    // initial_prompt: Optional[str]
    //        Optional text to provide as a prompt for the first window. This can be used to provide, or
    //        "prompt-engineer" a context for transcription, e.g. custom vocabularies or proper nouns
    //        to make it more likely to predict those word correctly.
    //params.initial_prompt = "Transcription of Tactical Combat Casualty Drugs such as Fentanyl, Ibuprofen, Amoxicillin, Epinephrine, TXA, Hextend, Ketamine, Oral Transmucosal Fentanyl Citrate. ";
    params.offset_ms = 0;
    params.no_context = true;
    params.single_segment   = true; //hard code for true, objc example has it based on a button press
    params.no_timestamps    = params.single_segment; //from streaming objc example

    whisper_reset_timings(context);

    LOGI("About to run whisper_full");
    if (whisper_full(context, params, audio_data_arr, audio_data_length) != 0) {
        LOGI("Failed to run the model");
    } else {
        whisper_print_timings(context);
    }
    (*env)->ReleaseFloatArrayElements(env, audio_data, audio_data_arr, JNI_ABORT);
}
```
Thanks so much.

**Comment by zhouwg** [comment]
ASR/transcription performance on Xiaomi 14 is about **5x-20x** better then other Android phone(low-end phone from vivo, huawei), but it's still **not enough** for real-time subtitle with online TV.


![1928721132](https://github.com/ggerganov/whisper.cpp/assets/6889919/6713b06c-93d7-4c6f-9afe-f41b72cc92e0)

![48922803](https://github.com/ggerganov/whisper.cpp/assets/6889919/23012f1b-81c6-4f65-ac71-60f8f2feb66a)

transcription performance can be improved by about **1-3 seconds when enable openblas**.

performance of mulmat benchmark seems be improved a lot / **significantly when enable openblas**.

so I guess Apple's dedicated machine learning acceleration library might be very important for performance on iOS/Mac.just like Georgi Gerganov said before. and we should/might study something about **Qualcomm's dedicated machine learning acceleration library** accordingly.


updated on 03-10-2024(2024-03-10, 15:46, Beijing Time / GMT + 8), 

![525154250](https://github.com/ggerganov/whisper.cpp/assets/6889919/3855adc8-e0d1-4860-a1bd-c840fea4c319)


![785005777](https://github.com/ggerganov/whisper.cpp/assets/6889919/16ea2038-ca86-4e8b-9a63-ee63349432d9)

updated on 03-10-2024(2024-03-10, 23:41 Beijing Time / GMT + 8):

![Screenshot from 2024-03-10 23-17-28](https://github.com/ggerganov/whisper.cpp/assets/6889919/62471964-b1e9-4ca0-843a-0b3d3b914e6a)
![1228365447](https://github.com/ggerganov/whisper.cpp/assets/6889919/e3e64204-7622-4c82-a57e-6376cb6df669)


I think I got the point although it's still not enough for real-time subtitle with online TV.

updated on 03-11-2024(2024-03-11,10:40) Beijing Time / GMT + 8

![1210108450](https://github.com/cdeos/kantv/assets/6889919/bdaa5d3e-9f4b-43a2-9626-6f388e9cbdcb)

less than 2 seconds for the first time.

commit could be found [here](https://github.com/cdeos/kantv/commit/d39eb0a9cfb258c5327ea65627662bb7dd88f81b). FYI.



updated on 03-12-2024(2024-03-12,00:51 , Beijing Time / GMT +8)
(12:45 means 00:45 in 03-12-2024)

![1924953245](https://github.com/cdeos/kantv/assets/6889919/773a2910-ce21-46a1-81c2-1aff63663723)
![Screenshot from 2024-03-12 00-46-14](https://github.com/cdeos/kantv/assets/6889919/e33c6178-b882-485d-acbb-338a1402694e)

less than 1 second for the first time.

AI is really a magic technology. thanks for great whisper.cpp.

<hr>

there is a **very concise** implementation of Android example during [my on-going PoC](https://github.com/cdeos/kantv/issues/64).

some source codes could be found here:


https://github.com/cdeos/kantv/blob/kantv-poc-with-whispercpp/external/whispercpp/whisper.cpp#L6726

https://github.com/cdeos/kantv/blob/kantv-poc-with-whispercpp/external/whispercpp/whisper.h#L620

https://github.com/cdeos/kantv/blob/kantv-poc-with-whispercpp/external/whispercpp/jni/whispercpp-jni.c

https://github.com/cdeos/kantv/blob/kantv-poc-with-whispercpp/cdeosplayer/cdeosplayer-lib/src/main/java/org/ggml/whispercpp/whispercpp.java

https://github.com/cdeos/kantv/blob/kantv-poc-with-whispercpp/cdeosplayer/kantv/src/main/java/com/cdeos/kantv/ui/fragment/ASRResearchFragment.java


if it's considered well and accepted, I'd like to spent time(clean up codes and remove any dependence......, just follow whisper.cpp's principle: without dependencies, keep clean, ...) to merge code from there to here and submit a PR accordingly. 
if not, I'll skip it.
thanks so much.

**Comment by zhouwg** [comment]
updated on 03-16-2024,22:36 (Beijing Time / GMT + 8),  

real-time English subtitle for online TV by whisper.cpp


https://github.com/cdeos/kantv/assets/6889919/7f571d30-c77c-4897-8cbd-53268284e641

**Reply by vishnusureshperumbavoor** [reply]
how to do this

**Comment by zhouwg** [comment]
updated on 03-24-2024, 22:55,

@ggerganov, better performance with better stability after finetune with method which introduced in https://github.com/ggerganov/whisper.cpp/issues/1951, but sometimes whisper.cpp will produce meaningless repeat tokens.


BTW, the root cause of the crash was that there were some invalid UTF-8 strings coming from whisper.cpp.

https://stackoverflow.com/questions/60722231/jni-detected-error-in-application-input-is-not-valid-modified-utf-8-illegal-st

https://github.com/zhouwg/kantv/pull/97


https://github.com/zhouwg/kantv/assets/6889919/2fabcb24-c00b-4289-a06e-05b98ecd22b8
